core/src/cell.rs

```
@@ -168,65 +136,69 @@ impl<'a> CellProvider for BlockCellProvider<'a> {
-        .collect();
+            };
 
-    let dep_cells = transaction
-        .dep_pts()
-        .iter()
-        .map(|dep| {
-            if seen_inputs.contains(dep) {
-                CellStatus::Dead
-            } else {
-                cell_provider.get_cell_status(dep)
+            match cell_status {
+                CellStatus::Dead => dead_out_points.push(out_point.clone()),
+                CellStatus::Unknown => unknown_out_points.push(out_point.clone()),
+                CellStatus::Live(cell_meta) => input_cells.push(cell_meta),
             }
         })
-        .collect();
-
-    ResolvedTransaction {
-        transaction,
-        input_cells,
-        dep_cells,
     }
-}
 
-impl<'a> ResolvedTransaction<'a> {
-    pub fn cells_iter(&self) -> Chain<slice::Iter<CellStatus>, slice::Iter<CellStatus>> {
-        self.dep_cells.iter().chain(&self.input_cells)
-    }
-
-    pub fn cells_iter_mut(
-        &mut self,
-    ) -> Chain<slice::IterMut<CellStatus>, slice::IterMut<CellStatus>> {
-        self.dep_cells.iter_mut().chain(&mut self.input_cells)
-    }
-
-    pub fn is_double_spend(&self) -> bool {
-        self.cells_iter().any(CellStatus::is_dead)
-    }
-
-    pub fn is_orphan(&self) -> bool {
-        self.cells_iter().any(CellStatus::is_unknown)
-    }
+    transaction.dep_pts().iter().for_each(|out_point| {
+        let cell_status = if seen_inputs.contains(out_point) {
+            CellStatus::Dead
+        } else {
+            cell_provider.cell(out_point)
+        };
 
-    pub fn is_fully_resolved(&self) -> bool {
-        self.cells_iter().all(CellStatus::is_live)
+        match cell_status {
+            CellStatus::Dead => dead_out_points.push(out_point.clone()),
+            CellStatus::Unknown => unknown_out_points.push(out_point.clone()),
+            CellStatus::Live(cell_meta) => dep_cells.push(cell_meta),
+        }
+    });
+
+    if !dead_out_points.is_empty() {
+        Err(UnresolvableError::Dead(dead_out_points))
```

Do we really need to save all failed resolved out_points? why not return an error on the first failure?

What happened if we have both Dead and Unknown resolves?

---

We can early return `Dead` as a short circuit error, but cannot do for `Unknown`, which is used in the tx pool:
https://github.com/nervosnetwork/ckb/blob/cc16678426b76309967f4c8880a4fc9368434741/shared/src/chain_state.rs#L242

---

`Dead` is a higher priority error than `Unknown`, however, to make code easy to understand and debug, I choose to return all error `OutPoint` instead of first one.

---

IMO short-cut return Dead error is more efficient and clear.

---

I think it's better to return error asap in `resolve_transaction`.